### PR TITLE
Config to allow subset of processors and exporters

### PR DESCRIPTION
**Pull request by oppegard**
This adds two new job properties that let's an operator restrict the kinds of OTel processors and exporters that can be used: 
`allow_list.processors` and `allow_list.exporters`. These properties are empty by default (`[]`), which keeps the existing behavior of allowing all processors and exporters that are included in our distribution. If not empty, then only the included components are allowed. For example, given `allow_list.exporters: [otlp]` and a `config` that contain the `prometheus` exporter, an error will be raised during template rendering of `config.yml.erb`.

```
Task 202 | 20:25:55 | Preparing deployment: Rendering templates (00:00:18)
                    L Error: Unable to render instance groups for deployment. Errors are:
  - Unable to render jobs for instance group 'blobstore'. Errors are:
    - Unable to render templates for job 'otel-collector'. Errors are:
      - Error filling in template 'config.yml.erb' (line 130: The following configured exporters are not allowed: ["prometheus"]. Allowed: ["otlp"].)
```

Given the non-GA status of the majority of OTel components, using these new properties helps limit which of the components can be used by end-users (in our case, only the OTel Collector configuration yaml of `config` is exposed to end-users, and we can set the new allow list properties).

We also added the `memory_limiter` processor in our distribution.

### CODE REVIEW

**Review by ctlong — COMMENTED** [review]
Overall LGTM! Just a few questions and concerns.

**Review by mkocher — COMMENTED** [review]
Left a few minor comments. Let's also rebase the commits and clean up the commit messages before merging this PR

File: spec/support/shared_examples_for_otel_collector.rb [hunk 1]
```diff
@@ -182,26 +182,123 @@ def without_internal_telemetry(cfg)
...
+          expect(error.message).to include("Available: [#{formatted_names}]")
+        end
+      end
+
       it 'includes the configured processors in the config' do
         expect(rendered.keys).to include 'processors'
         expect(rendered['processors']).to eq(config['processors'])
       end
+
+      it 'includes the configured processors even if their names contain `/`' do
```
[the remark this answers is not in the document]
Comment (oppegard) [reply spec/support/shared_examples_for_otel_collector.rb:212]: We wanted to specifically test the use of `split` in the following line:

```
user_components = config.fetch(component_kind, {}).map {|key, _| key.split('/')[0]}
```

File: jobs/otel-collector-windows/spec [hunk 2]
```diff
@@ -40,6 +40,14 @@ properties:
             receivers: [otlp/placeholder]
             processors: [batch]
             exporters: [otlp]
+  processorsAllowList:
+    description: "Processors allowed for use in otel-collector config. Must be a subset of list included in otel-collector builder config. Empty list means allow all possible."
+    default: []
+    example: ['batch']
+  exportersAllowList:
```
[the remark this answers is not in the document]
Comment (oppegard) [reply jobs/otel-collector-windows/spec:47]: A deny list seems a bit more brittle than an allow list. For an operator deploying this BOSH release, I can envision setting an allow list based on the current exporters and processors included in the release. However, if a new version of the release is cut that includes new components, an operator would have to update their deny lists to exclude them. It seems less brittle to just set `allow_list.exporters: [otlp]` and expect that to remain working as new versions of the BOSH release are consumed.

**Review by ctlong — APPROVED** [review]

**Pull request merged**